### PR TITLE
Add bandit class

### PR DESCRIPTION
**Pull request by guidopetri**
First version of the bandit class. It's using policy gradient to update the `theta` params which are actually serving as a weighted sum of the context, *not* as the raw preference for each action like in the slides. Most probably, we'll call this with "epochs" or via random sampling of the context/action/reward pairs we have in our data.

I'd greatly appreciate a code review so we know I was not sleep deprived when I wrote this. (I was.)

Usage:

```python
bandit = Bandit(n_context_features, n_arms)

# get action probabilities
bandit.predict_proba(X_context[:2])

# get actions with highest probability
bandit.predict(X_context[:2], deterministic=True)

# get actions via random sampling
bandit.predict(X_context[:2], deterministic=False)

# update theta parameters
bandit.update_theta(X_context[:2], X_action[:2], Y_reward[:2])

# also available: baseline, currently just mean reward so far (not per-action)
bandit = Bandit(n_context_features, n_arms, baseline=True)
bandit.update_theta(X_context[:2], X_action[:2], Y_reward[:2])  # different theta than earlier

# get theta params
print(bandit.theta)
```

### CODE REVIEW

File: bandit.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        # start at uniform
+        self.theta = np.zeros((self.n_features, self.n_arms))
+        
+    @property
+    def current_baseline(self):
+        return (self.reward_sum / self.iters
+                if self.baseline and self.iters
+                else 0)
+        
+    def predict_proba(self, X):
```
Comment (mihamerstan) [inline bandit.py:26]: I think this is fine for now (and I see where it came from in the Policy Gradient slides), but isn't this restricting our policy space to softmax-on-linear models? We'll probably want to be able to consider other policy functions down the road.

File: bandit.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        return softmax((X @ self.theta).reshape(-1, self.n_arms), axis=1)
+    
+    def predict(self, X, deterministic=True):
+        predictions = self.predict_proba(X)
+        if deterministic:
+            return predictions.argmax()
+        cumsum = predictions.cumsum(axis=1)
+        random_val = np.random.random_sample(len(cumsum))
+        binarized = (cumsum.T < random_val).astype(int).sum(axis=0)
```
Comment (mihamerstan) [inline bandit.py:35]: whoa, that's fancy. If you want to be less fancy next time, np.random.choice() is pretty sweet and draws randomly from a distribution.

File: bandit.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                if self.baseline and self.iters
+                else 0)
+        
+    def predict_proba(self, X):
+        return softmax((X @ self.theta).reshape(-1, self.n_arms), axis=1)
+    
+    def predict(self, X, deterministic=True):
+        predictions = self.predict_proba(X)
+        if deterministic:
+            return predictions.argmax()
```
Comment (mihamerstan) [inline bandit.py]: this should be predictions.argmax(axis=1) so you get a vector

**Review by mihamerstan — COMMENTED** [review]
I left a few comments in the code, but otherwise it looks good.

File: bandit.py (hunk 1 above)
Comment (guidopetri) [reply bandit.py:26]: it is, yes. nonlinear transforms can still be done (we just have to do them manually), but we still require softmax no matter what. not sure what other policies we might pick without softmax that we can train via policy gradient

File: bandit.py (hunk 3 above)
Comment (guidopetri) [reply bandit.py]: Nice catch, thanks!

File: bandit.py (hunk 2 above)
Comment (guidopetri) [reply bandit.py:35]: I tried that - unfortunately it doesn't work for different distributions for each row (which is what we have). Specifically, the `p` parameter needs to be a vector/1D array, not a matrix.

Comment (guidopetri) [reply bandit.py:35]: (if you have a better way of doing this - I agree it looks very odd and takes some mental gymnastics - then please, let me know, haha)
